In [1]:
import pandas as pd
import logging

# Log fayl manzili
log_path = r"C:\Users\Rasulbek907\Desktop\SML_2_Pr\Log\feature_selection.log"

# Log sozlamalari
logging.basicConfig(
    filename=log_path,
    filemode='a',  # Append mode
    format='%(asctime)s - %(levelname)s - %(message)s',
    level=logging.INFO
)

try:
    logging.info("CSV fayl o'qilmoqda: data_merged.csv")
    df = pd.read_csv(r"C:\Users\Rasulbek907\Desktop\SML_2_Pr\Data\data_preprocessing\Preprocessed.csv")
    logging.info(f"Fayl muvaffaqiyatli o'qildi. Satırlar soni: {len(df)} ustunlar soni: {len(df.columns)}")
except Exception as e:
    logging.error(f"CSV faylni o'qishda xatolik: {e}")


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 489121 entries, 0 to 489120
Data columns (total 13 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   Unnamed: 0.1           489121 non-null  float64
 1   Unnamed: 0             489121 non-null  float64
 2   id                     489121 non-null  float64
 3   week                   489121 non-null  float64
 4   center_id              489121 non-null  float64
 5   meal_id                489121 non-null  float64
 6   checkout_price         489121 non-null  float64
 7   base_price             489121 non-null  float64
 8   emailer_for_promotion  489121 non-null  float64
 9   homepage_featured      489121 non-null  float64
 10  num_orders             489121 non-null  float64
 11  category               489121 non-null  float64
 12  cuisine                489121 non-null  float64
dtypes: float64(13)
memory usage: 48.5 MB


In [3]:
df.head()

Unnamed: 0.1  Unnamed: 0        id      week  center_id   meal_id  \
0     -1.732047   -1.732047  0.897671 -1.777284  -0.589303 -0.255481   
1     -1.732040   -1.732040  1.503205 -1.777284  -0.589303 -0.058187   
2     -1.732033   -1.732033  0.672019 -1.777284  -0.589303  0.939240   
3     -1.732026   -1.732026  0.611351 -1.777284  -0.589303  0.208524   
4     -1.732019   -1.732019  1.375217 -1.777284  -0.589303  1.107304   

   checkout_price  base_price  emailer_for_promotion  homepage_featured  \
0          136.83      152.29              -0.295229          -0.346777   
1          136.83      135.83              -0.295229          -0.346777   
2          134.86      135.86              -0.295229          -0.346777   
3          339.50      437.53              -0.295229          -0.346777   
4          243.50      242.50              -0.295229          -0.346777   

   num_orders  category   cuisine  
0       177.0 -0.413866 -1.450253  
1       270.0  0.604420 -0.287105  
2       189.0  0.095277 -0.287105  
3        54.0  1.622706  0.876043  
4        40.0  1.113563 -0.868679

In [5]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()

cols_to_scale = ['checkout_price', 'base_price', 'num_orders']
df[cols_to_scale] = scaler.fit_transform(df[cols_to_scale])


In [6]:
df.head()

Unnamed: 0.1  Unnamed: 0        id      week  center_id   meal_id  \
0     -1.732047   -1.732047  0.897671 -1.777284  -0.589303 -0.255481   
1     -1.732040   -1.732040  1.503205 -1.777284  -0.589303 -0.058187   
2     -1.732033   -1.732033  0.672019 -1.777284  -0.589303  0.939240   
3     -1.732026   -1.732026  0.611351 -1.777284  -0.589303  0.208524   
4     -1.732019   -1.732019  1.375217 -1.777284  -0.589303  1.107304   

   checkout_price  base_price  emailer_for_promotion  homepage_featured  \
0        0.120524    0.091689              -0.295229          -0.346777   
1        0.120524    0.076121              -0.295229          -0.346777   
2        0.118750    0.076149              -0.295229          -0.346777   
3        0.303003    0.361478              -0.295229          -0.346777   
4        0.216567    0.177012              -0.295229          -0.346777   

   num_orders  category   cuisine  
0    0.006753 -0.413866 -1.450253  
1    0.010582  0.604420 -0.287105  
2    0.007247  0.095277 -0.287105  
3    0.001688  1.622706  0.876043  
4    0.001112  1.113563 -0.868679

In [7]:
columns_list = df.columns.tolist()
print(columns_list)


['Unnamed: 0.1', 'Unnamed: 0', 'id', 'week', 'center_id', 'meal_id', 'checkout_price', 'base_price', 'emailer_for_promotion', 'homepage_featured', 'num_orders', 'category', 'cuisine']


# Narxlar orasidagi farqni hisoblash

In [8]:
df['price_diff'] = df['base_price'] - df['checkout_price']
df['price_diff_ratio'] = df['price_diff'] / df['base_price']

# Haftaga oid yangi feature yaratish // week ustunidan foydalangan holda: Sezon (masalan, 4 haftalik davrlarni guruhlash), Oy ichidagi hafta kabi feature yaratish mumkin.

In [ ]:
df['season'] = (df['week'] // 4) + 1  # har 4 haftaga bitta sezon

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 489121 entries, 0 to 489120
Data columns (total 16 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   Unnamed: 0.1           489121 non-null  float64
 1   Unnamed: 0             489121 non-null  float64
 2   id                     489121 non-null  float64
 3   week                   489121 non-null  float64
 4   center_id              489121 non-null  float64
 5   meal_id                489121 non-null  float64
 6   checkout_price         489121 non-null  float64
 7   base_price             489121 non-null  float64
 8   emailer_for_promotion  489121 non-null  float64
 9   homepage_featured      489121 non-null  float64
 10  num_orders             489121 non-null  float64
 11  category               489121 non-null  float64
 12  cuisine                489121 non-null  float64
 13  price_diff             489121 non-null  float64
 14  price_diff_ratio       489121 non-nu

In [11]:
skewness = df.skew(numeric_only=True)
print(skewness)


Unnamed: 0.1             0.000000
Unnamed: 0               0.000000
id                       0.000025
week                    -0.047665
center_id                0.345446
meal_id                 -0.172985
checkout_price           0.667832
base_price               0.639135
emailer_for_promotion    3.091980
homepage_featured        2.536927
num_orders               7.172919
category                 0.070537
cuisine                 -0.271418
price_diff               2.726384
price_diff_ratio              NaN
season                  -0.029857
dtype: float64


c:\Users\Rasulbek907\Desktop\SML_2_Pr\myvenv\Lib\site-packages\pandas\core\nanops.py:1256: RuntimeWarning: invalid value encountered in subtract
  adjusted = values - mean


| Ustun                   | Skewness | Xulosa                                                           |
| ----------------------- | -------- | ---------------------------------------------------------------- |
| `num_orders`            | **7.17** | Juda o‘ngga og‘gan → ⚠️ Log transformatsiya tavsiya qilinadi     |
| `emailer_for_promotion` | **3.09** | Juda qiyshiq (bool ustun) → OK, lekin ehtiyot bo‘lish kerak      |
| `homepage_featured`     | **2.53** | Shuningdek bool ustun — ayniqsa kam hollarda `1` bo‘lsa          |
| `price_diff`            | **2.72** | O‘ngga og‘gan → Log yoki sqrt transformatsiya foydali            |
| `checkout_price`        | **0.66** | Biroz og‘gan — mumkin bo‘lsa yengil transformatsiya              |
| `base_price`            | **0.63** | Shunga o‘xshash                                                  |
| `price_diff_ratio`      | NaN      | Balki 0 yoki `base_price = 0` bo‘lgan joy bor — tekshirish kerak |


# Raqamli va skewness > 1 bo‘lgan ustunlarga fokus qilamiz:

In [12]:
import numpy as np

# log1p = log(x + 1), bu 0 ga teng bo‘lgan qiymatlar uchun xavfsizroq
df['num_orders_log'] = np.log1p(df['num_orders'])
df['price_diff_log'] = np.log1p(df['price_diff'])

# Optional: checkout_price, base_price
df['checkout_price_log'] = np.log1p(df['checkout_price'])
df['base_price_log'] = np.log1p(df['base_price'])


# price_diff_ratio ustunidagi NaN muammo

In [13]:
# Avval base_price = 0 bo'lganlarni tekshiramiz
print(df[df['base_price'] == 0])

# Ehtimol, bu yerda bo'sh yoki xatolik bor — buni ehtiyotkorlik bilan hal qilish kerak
# Masalan, price_diff_ratio ustunini qayta hisoblaymiz:
df['price_diff_ratio'] = df['price_diff'] / (df['base_price'] + 1e-5)


        Unnamed: 0.1  Unnamed: 0        id      week  center_id   meal_id  \
157699     -0.615176   -0.615176  0.665015 -0.604615  -0.045509 -0.405277   

        checkout_price  base_price  emailer_for_promotion  homepage_featured  \
157699        0.047162         0.0              -0.295229          -0.346777   

        num_orders  category   cuisine  price_diff  price_diff_ratio  season  \
157699    0.002841 -0.413866 -0.868679   -0.047162              -inf     0.0   

        num_orders_log  price_diff_log  checkout_price_log  base_price_log  
157699        0.002837        -0.04831            0.046083             0.0  


# Skewnessdan keyin yana normalizatsiya (scaling)

In [14]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
scaled_cols = ['num_orders_log', 'price_diff_log', 'checkout_price_log', 'base_price_log']
df[scaled_cols] = scaler.fit_transform(df[scaled_cols])


In [15]:
skewness = df.skew(numeric_only=True)
print(skewness)

Unnamed: 0.1               0.000000
Unnamed: 0                 0.000000
id                         0.000025
week                      -0.047665
center_id                  0.345446
meal_id                   -0.172985
checkout_price             0.667832
base_price                 0.639135
emailer_for_promotion      3.091980
homepage_featured          2.536927
num_orders                 7.172919
category                   0.070537
cuisine                   -0.271418
price_diff                 2.726384
price_diff_ratio        -698.233455
season                    -0.029857
num_orders_log             5.911747
price_diff_log             2.558554
checkout_price_log         0.467729
base_price_log             0.444863
dtype: float64


In [16]:
df['num_orders_sqrt'] = np.sqrt(df['num_orders'])


# price_diff_ratio ustuni: Skewness -698 bo‘lishi — bu yerda ehtimol base_price = 0 bo‘lib, katta manfiy qiymatlar kelib chiqqan.

In [17]:
# Avval bazaviy 0 bo'lganlar sonini ko‘raylik
print(df[df['base_price'] == 0].shape[0])

# So'ng NaN yoki 0 bo'lishdan qochish uchun denominatorga kichik son qo‘shamiz:
df['price_diff_ratio'] = df['price_diff'] / (df['base_price'] + 1e-5)

# Yoki: limit qo‘yamiz
df['price_diff_ratio'] = df['price_diff_ratio'].clip(-10, 10)


1


In [18]:
skewness = df.skew(numeric_only=True)
print(skewness)

Unnamed: 0.1             0.000000
Unnamed: 0               0.000000
id                       0.000025
week                    -0.047665
center_id                0.345446
meal_id                 -0.172985
checkout_price           0.667832
base_price               0.639135
emailer_for_promotion    3.091980
homepage_featured        2.536927
num_orders               7.172919
category                 0.070537
cuisine                 -0.271418
price_diff               2.726384
price_diff_ratio        -1.177564
season                  -0.029857
num_orders_log           5.911747
price_diff_log           2.558554
checkout_price_log       0.467729
base_price_log           0.444863
num_orders_sqrt          1.429111
dtype: float64
